src/modules/rtp_media_server/rtp_media_server.c

```diff
@@ -0,0 +1,716 @@
+	}
+	ortp_set_log_file(log_file);
+	ortp_set_log_level_mask(
+			NULL, ORTP_MESSAGE | ORTP_WARNING | ORTP_ERROR | ORTP_FATAL);
+	return (0);
+}
+
+/**
+ * Called only once when OpenSER is shuting down to clean up module
+ * resources.
+ */
+static void mod_destroy()
+{
+	rms_media_destroy();
+	LM_INFO("RTP media server module destroy\n");
+	return;
+}
+
+void rms_signal_handler(int signum)
+{
+	LM_INFO("signal received [%d]\n", signum);
+}
+
+/**
+ * Most interaction with the session and media streams that are controlled 
+ * in this function this is safer in the event where a library is using non shared memory
+ * all the mediastreamer2 ticker threads are spawned from here.
+ */
+static void rms_session_manage_loop()
+{
+	while(1) {
+		lock(&session_list_mutex);
+		rms_session_info_t *si;
+		clist_foreach(rms_session_list, si, next)
+		{
+			if(si->action == RMS_HANGUP) {
+				LM_INFO("session action RMS_HANGUP [%s]\n", si->callid.s);
+				rms_hangup_call(si);
+				si->action = RMS_STOP;
+			} else if(si->action == RMS_STOP) {
+				LM_INFO("session action RMS_STOP [%s]\n", si->callid.s);
+				rms_stop_media(&si->caller_media);
+				rms_session_info_t *tmp = si->prev;
+				rms_session_free(si);
+				si = tmp;
+			} else if(si->action == RMS_PLAY) {
+				LM_INFO("session action RMS_PLAY [%s]\n", si->callid.s);
+				rms_playfile(&si->caller_media, si->action_param.s);
+				si->action = RMS_NONE;
+			} else if(si->action == RMS_DONE) {
+				LM_INFO("session action RMS_DONE [%s][%s]\n", si->callid.s,
+						si->action_route.s);
+				if(si->action_route.s) {
+					run_action_route(si, si->action_route.s);
+				} else {
+					si->action = RMS_HANGUP;
+				}
+			} else if(si->action == RMS_START) {
+				create_call_leg_media(&si->caller_media);
+				LM_INFO("session action RMS_START [%s]\n", si->callid.s);
+				rms_start_media(&si->caller_media, si->action_param.s);
+				si->action = RMS_NONE;
+				run_action_route(si, "rms:start");
+			}
+		}
+		unlock(&session_list_mutex);
+		usleep(2000);
+	}
+}
+
+/**
+ * The rank will be o for the main process calling this function,
+ * or 1 through n for each listener process. The rank can have a negative
+ * value if it is a special process calling the child init function.
+ * Other then the listeners, the rank will equal one of these values:
+ * PROC_MAIN      0  Main ser process
+ * PROC_TIMER    -1  Timer attendant process
+ * PROC_FIFO     -2  FIFO attendant process
+ * PROC_TCP_MAIN -4  TCP main process
+ * PROC_UNIXSOCK -5  Unix domain socket server processes
+ *
+ * If this function returns a nonzero value the loading of OpenSER will
+ * stop.
+ */
+static int child_init(int rank)
+{
+	if(rank == PROC_MAIN) {
+		int pid;
+		pid = fork_process(PROC_XWORKER, "RTP_media_server", 1);
+		if(pid < 0)
+			return -1;
+		if(pid == 0) {
+			rms_session_manage_loop();
+			return 0;
+		}
+	}
+	int rtn = 0;
+	return (rtn);
+}
+
+int rms_str_dup(str *dst, str *src, int shared)
```

Comment:
please use the pkg_str_dup() and shm_str_dup() in the core

Reply:
Same topic as before, since they do not convert to `\0` terminated `char *` they are not flexible in this case.